curvefs/src/metaserver/metastore.cpp

```diff
@@ -91,14 +139,45 @@ void MetaStoreImpl::SaveBackground(const std::string& path,
     done->Run();
 }
 
-bool MetaStoreImpl::Save(const std::string& path,
+bool MetaStoreImpl::Save(const std::string& dir,
                          OnSnapshotSaveDoneClosure* done) {
+    brpc::ClosureGuard doneGuard(done);
     WriteLockGuard writeLockGuard(rwLock_);
-    DumpFileClosure child;
-    std::thread th = std::thread(&MetaStoreImpl::SaveBackground,
-                                 this, path, &child, done);
-    child.WaitRunned();
-    th.detach();
+
+    MetaStoreFStream fstream(&partitionMap_, kvStorage_,
+                             copysetNode_->GetPoolId(),
+                             copysetNode_->GetCopysetId());
+
+    const std::string metadata = dir + "/" + kMetaDataFilename;
+    bool succ = fstream.Save(metadata);
+    if (!succ) {
+        done->SetError(MetaStatusCode::SAVE_META_FAIL);
+        return false;
+    }
+
+    // checkpoint storage
+    butil::Timer timer;
+    timer.start();
+    std::vector<std::string> files;
+    succ = kvStorage_->Checkpoint(dir, &files);
```

Comment:
Here, checkpoint only needs to call `kvStorage_->Checkpoint(dir, &files)`. 

Why does it need an empty read-only db to operate when restoring from sanpshot `uplicateRocksdbCheckpoint(dir + "/" + kRocksdbCheckpointPath, options_.dataDir);`? 

Is there no interface to restore directly to the checkpoint point?

Reply:
the checkpoint is in raft snapshot, we must not modify it, so when restoring from the snapshot, we had to copy the checkpoint and open it, and the efficient way is checkpoint again from the checkpoint in snapshot.